components/scream/src/share/grid/remap/coarsening_remapper.cpp

```diff
@@ -0,0 +1,965 @@
+
+  const int nlevs  = src_grid->get_num_vertical_levels();
+
+  auto tgt_grid_gids = m_ov_tgt_grid->get_unique_gids ();
+  auto tgt_grid = std::make_shared<PointGrid>("tgt_grid",tgt_grid_gids.size(),nlevs,m_comm);
+  tgt_grid->set_dofs(tgt_grid_gids);
+  this->set_grids(src_grid,tgt_grid);
+}
+
+CoarseningRemapper::
+~CoarseningRemapper ()
+{
+  // We need to free MPI requests
+  for (size_t i=0; i<m_send_req.size(); ++i) {
+    MPI_Request_free(&m_send_req[i]);
+  }
+  for (size_t i=0; i<m_recv_req.size(); ++i) {
+    MPI_Request_free(&m_recv_req[i]);
+  }
+}
+
+FieldLayout CoarseningRemapper::
+create_src_layout (const FieldLayout& tgt_layout) const
+{
+  using namespace ShortFieldTagsNames;
+  const auto lt = get_layout_type(tgt_layout.tags());
+  auto src = FieldLayout::invalid();
+  const bool midpoints = tgt_layout.has_tag(LEV);
+  const int vec_dim = tgt_layout.is_vector_layout() ? tgt_layout.get_vector_dim() : -1;
+  switch (lt) {
+    case LayoutType::Scalar2D:
+      src = m_src_grid->get_2d_scalar_layout();
+      break;
+    case LayoutType::Vector2D:
+      src = m_src_grid->get_2d_vector_layout(CMP,vec_dim);
+      break;
+    case LayoutType::Scalar3D:
+      src = m_src_grid->get_3d_scalar_layout(midpoints);
+      break;
+    case LayoutType::Vector3D:
+      src = m_src_grid->get_3d_vector_layout(midpoints,CMP,vec_dim);
+      break;
+    default:
+      EKAT_ERROR_MSG ("Layout not supported by CoarseningRemapper: " + e2str(lt) + "\n");
+  }
+  return src;
+}
+FieldLayout CoarseningRemapper::
+create_tgt_layout (const FieldLayout& src_layout) const
+{
+  using namespace ShortFieldTagsNames;
+  const auto lt = get_layout_type(src_layout.tags());
+  auto tgt = FieldLayout::invalid();
+  const bool midpoints = src_layout.has_tag(LEV);
+  const int vec_dim = src_layout.is_vector_layout() ? src_layout.get_vector_dim() : -1;
+  switch (lt) {
+    case LayoutType::Scalar2D:
+      tgt = m_tgt_grid->get_2d_scalar_layout();
+      break;
+    case LayoutType::Vector2D:
+      tgt = m_tgt_grid->get_2d_vector_layout(CMP,vec_dim);
+      break;
+    case LayoutType::Scalar3D:
+      tgt = m_tgt_grid->get_3d_scalar_layout(midpoints);
+      break;
+    case LayoutType::Vector3D:
+      tgt = m_tgt_grid->get_3d_vector_layout(midpoints,CMP,vec_dim);
+      break;
+    default:
+      EKAT_ERROR_MSG ("Layout not supported by CoarseningRemapper: " + e2str(lt) + "\n");
+  }
+  return tgt;
+}
+
+void CoarseningRemapper::
+do_register_field (const identifier_type& src, const identifier_type& tgt)
+{
+  m_src_fields.push_back(field_type(src));
+  m_tgt_fields.push_back(field_type(tgt));
+}
+
+void CoarseningRemapper::
+do_bind_field (const int ifield, const field_type& src, const field_type& tgt)
+{
+  EKAT_REQUIRE_MSG (
+      src.get_header().get_identifier().get_layout().rank()>1 ||
+      src.get_header().get_alloc_properties().get_padding()==0,
+      "Error! We don't support 2d scalar fields that are padded.\n");
```

Comment:
We support rank 3 field layouts, too, right? Maybe "2d/3d" instead of "2d" here?

Reply:
Gotcha. Thanks for setting me straight.